_posts/2023-06-19-End-to-end-acknowledgements-in-Data-Prepper.md

```diff
@@ -0,0 +1,64 @@
+---
+layout: post
+title:  "End-to-end acknowledgements in Data Prepper"
+authors:
+- kkondaka
+- dlv
+date: 2023-06-19 10:00:00 -0700
+categories:
+  - technical-post
+
+meta_keywords: Data Prepper E2E, Data Prepper sink, Data Prepper pipeline as the sink
+meta_description: Learn how Data Prepper provides data durability through the use of end to end acknowledgements by delivering data to the sink before notifying the source.
+
+---
+
+[Data Prepper](https://opensearch.org/docs/latest/data-prepper/index/) is an open-source data collector used to ingest data into OpenSearch clusters. While Data Prepper has an in-memory buffer that allows for fast throughput, users also want improved data durability, particularly, confirmation that the data received by Data Prepper reaches the desired sink. To solve challenges with data durability, Data Prepper now has capability to use end-to-end acknowledgements.
+
+## The need: Improving data durability 
+
+The Data Prepper maintainers and their teams have observed that a common challenge to data durability is the reliability of the ingestion pipeline. For example, if the OpenSearch cluster cannot receive data because of temporary stress on the cluster or underscaling, then Data Prepper cannot send data to the destination sink. Further, Data Prepper may exhaust memory or other system resources or the hardware running Data Prepper may fail resulting the data loss during the ingestion.
+
+To solve these challenges, we must consider the data sources themselves. For example, the Amazon S3 source can read data from a highly durable store. When observing this, we realized that if Data Prepper fails to deliver data to OpenSearch, Data Prepper can retry reading data from Amazon S3. This, however, requires knowing when the data is written before deleting the SQS message that notifies Data Prepper of an available S3 object to consume.
+
+
+## Our solution: End-to-end acknowledgements
+
+Data Prepper provides data durability through the use of end-to-end acknowledgements. When a Data Prepper source is configured to use end-to-end acknowledgements, the source is notified only when the data is successfully delivered to the sink. If the source receives end-to-end acknowledgments, it can take appropriate actions such as removing SQS messages that have been successfully delivered to the sink or incrementing the commit offset in case of Kafka source. If end-to-end acknowledgements are not received, the source may retry operation or notify the external source of the failure. The following picture illustrates the control flow when a Data Prepper source is configured to use end-to-end acknowledgements.
```

Comment:
I am assuming you did not mean to replace "acknowledgments" with "acknowledgements" because we have agreed to use the spelling "acknowledgments" everywhere in this blog.